File: go/06-memoria/03-episodica/episodic_log.go
// Memoria episódica sin decay — log plano para prototipado.
// Variante mínima: solo append y búsqueda por texto o sesión.

// Cómo ejecutar: make go FILE=go/06-memoria/03-episodica/episodic_log.go

package main

import (
	"fmt"
	"strings"
	"time"
)

type EntradaLog struct {
	Contenido string
	SesionID  string
	Timestamp time.Time
}

type MemoriaLog struct {
	log []EntradaLog
}

func (m *MemoriaLog) Append(contenido, sesionID string) EntradaLog {
	e := EntradaLog{Contenido: contenido, SesionID: sesionID, Timestamp: time.Now()}
	m.log = append(m.log, e)
	return e
}

func (m *MemoriaLog) RecallRecent(n int) []EntradaLog {
	if len(m.log) <= n {
		return m.log
	}
	return m.log[len(m.log)-n:]
}

func (m *MemoriaLog) RecallSearch(query string, n int) []EntradaLog {
	q := strings.ToLower(query)
	var matches []EntradaLog
	for _, e := range m.log {
		if strings.Contains(strings.ToLower(e.Contenido), q) {
			matches = append(matches, e)
		}
	}
	if len(matches) <= n {
		return matches
	}
	return matches[len(matches)-n:]
}

func (m *MemoriaLog) RecallSession(sesionID string) []EntradaLog {
	var result []EntradaLog
	for _, e := range m.log {
		if e.SesionID == sesionID {
			result = append(result, e)
		}
	}
	return result
}

func (m *MemoriaLog) Len() int {
	return len(m.log)
}

func main() {
	mem := &MemoriaLog{}

	mem.Append("El usuario usa Python 3.12 en producción", "s1")
	mem.Append("Bug en auth.py línea 247: condición invertida", "s1")
	mem.Append("Decidimos usar PostgreSQL en lugar de SQLite", "s2")
	mem.Append("El módulo de billing tiene deuda técnica", "s2")
	mem.Append("Prefiere respuestas sin código cuando no es necesario", "s3")

	fmt.Printf("Total: %d episodios\n\n", mem.Len())

	fmt.Println("--- últimos 3 ---")
	for _, e := range mem.RecallRecent(3) {
		contenido := e.Contenido
		if len(contenido) > 60 {
			contenido = contenido[:60]
		}
		fmt.Printf("  [%s] %s\n", e.SesionID, contenido)
	}

	fmt.Println("\n--- búsqueda: 'usuario' ---")
	for _, e := range mem.RecallSearch("usuario", 5) {
		contenido := e.Contenido
		if len(contenido) > 60 {
			contenido = contenido[:60]
		}
		fmt.Printf("  [%s] %s\n", e.SesionID, contenido)
	}

	fmt.Println("\n--- sesión s1 ---")
	for _, e := range mem.RecallSession("s1") {
		contenido := e.Contenido
		if len(contenido) > 60 {
			contenido = contenido[:60]
		}
		fmt.Printf("  %s\n", contenido)
	}
}

File: typescript/06-memoria/03-episodica/episodic_log.ts
// Memoria episódica sin decay — log plano para prototipado.
// Variante mínima: solo append y búsqueda por texto o sesión.

// Cómo ejecutar: make ts SCRIPT=typescript/06-memoria/03-episodica/episodic_log.ts


interface EntradaLog {
  contenido: string;
  sesionId: string | null;
  timestamp: number;
}

class MemoriaLog {
  private _log: EntradaLog[] = [];

  append(contenido: string, sesionId: string | null = null): EntradaLog {
    const entrada: EntradaLog = { contenido, sesionId, timestamp: Date.now() };
    this._log.push(entrada);
    return entrada;
  }

  recallRecent(n: number = 5): EntradaLog[] {
    return this._log.slice(-n);
  }

  recallSearch(query: string, n: number = 5): EntradaLog[] {
    const q = query.toLowerCase();
    const matches = this._log.filter((e) => e.contenido.toLowerCase().includes(q));
    return matches.slice(-n);
  }

  recallSession(sesionId: string): EntradaLog[] {
    return this._log.filter((e) => e.sesionId === sesionId);
  }

  get length(): number {
    return this._log.length;
  }
}

const mem = new MemoriaLog();

mem.append("El usuario usa Python 3.12 en producción", "s1");
mem.append("Bug en auth.py línea 247: condición invertida", "s1");
mem.append("Decidimos usar PostgreSQL en lugar de SQLite", "s2");
mem.append("El módulo de billing tiene deuda técnica", "s2");
mem.append("Prefiere respuestas sin código cuando no es necesario", "s3");

console.log(`Total: ${mem.length} episodios\n`);

console.log("--- últimos 3 ---");
for (const e of mem.recallRecent(3)) {
  console.log(`  [${e.sesionId}] ${e.contenido.slice(0, 60)}`);
}

console.log("\n--- búsqueda: 'usuario' ---");
for (const e of mem.recallSearch("usuario")) {
  console.log(`  [${e.sesionId}] ${e.contenido.slice(0, 60)}`);
}

console.log("\n--- sesión s1 ---");
for (const e of mem.recallSession("s1")) {
  console.log(`  ${e.contenido.slice(0, 60)}`);
}
